subsys/bluetooth/controller/ll_sw/ull_adv_aux.c

```diff
@@ -411,8 +385,84 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
 
 #if defined(CONFIG_BT_CTLR_ADV_AUX_PDU_LINK)
 	if ((op == BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG) ||
-	    (op == BT_HCI_LE_EXT_ADV_OP_LAST_FRAG) ||
-	    ad_len_overflow) {
+	    (op == BT_HCI_LE_EXT_ADV_OP_LAST_FRAG)) {
+		/* in the previous step we duplicated the chain
+		 * the next step is to append new data in the last existing pdu in the chain,
+		 */
+
+		val_ptr = hdr_data;
+		*val_ptr++ = len;
+		(void)memcpy(val_ptr, &data, sizeof(data));
+
+		/* Append data to the last PDU */
+		chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+						      ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND,
+						      0U, hdr_data);
+
+		LL_ASSERT((!chain_err) || (chain_err == BT_HCI_ERR_PACKET_TOO_LONG));
+
+		/* FIXME: the code has become quite complex, an alternative and simpler
+		 * implementation would be to first fill an array with all data that
+		 * must be send, and create the chained PDUs from this array
+		 */
+		if (chain_err == BT_HCI_ERR_PACKET_TOO_LONG) {
+			/* We could not fit all the data, append as much as possible
+			 * ad_len_overflow_first_try is how much overflows without the AUX ptr
+			 * ad_len_overflow is how much overflows with the AUX ptr
+			 */
+			const uint16_t chain_add_fields = ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							  ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
+
+			val_ptr = hdr_data;
+			*val_ptr++ = len;
+			(void)memcpy(val_ptr, &data, sizeof(data));
+			val_ptr += sizeof(data);
+			*val_ptr++ = len;
+			(void)memcpy(val_ptr, &data, sizeof(data));
+			chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+							      chain_add_fields,
+							      0U, hdr_data);
+			ad_len_chain = hdr_data[ULL_ADV_HDR_DATA_AUX_PTR_PTR_OFFSET +
+						ULL_ADV_HDR_DATA_AUX_PTR_PTR_SIZE +
+						ULL_ADV_HDR_DATA_DATA_PTR_OFFSET +
+						ULL_ADV_HDR_DATA_DATA_PTR_SIZE];
+
+			/* len is the total amount of datawe want to add
+			 * ad_len_chain is the amount of data that does
+			 * not fit in the current PDU
+			 * the difference of the two is the amount that
+			 * we can fit in the current PDU
+			 */
+			ad_len = len - ad_len_chain;
```

Comment:
What happens if we have a completely full PDU that we want to add an aux ptr to (ie. we now have to move some existing advertising data to the next PDU)? This code doesn't seem to handle that case.

Reply:
That is handled by the ull_adv_aux_clear_set function. However this code is hard to follow, so we want to do a rewrite. See issue #63288 and #63290